src/bin.ts

```diff
@@ -178,75 +240,52 @@ const deploy = async ({ yes, bucket }: { yes: boolean, bucket: string }) => {
-            const key = createSafeS3Key(relative(publicDir, path));
-            const buffer = await readFile(path);
-            const tag = `"${createHash('md5').update(buffer).digest('hex')}"`;
-            const object = objects.find(object => object.Key === key && object.ETag === tag);
-
-            isKeyInUse[key] = true;
-
-            if (object) {
-                // object with exact hash already exists, abort.
-                return;
-            }
-
-            try {
-                const promise = new S3.ManagedUpload({
-                    service: s3,
-                    params: {
-                        Bucket: config.bucketName,
-                        Key: key,
-                        Body: fs.createReadStream(path),
-                        ACL: config.acl === null ? undefined : (config.acl || 'public-read'),
-                        ContentType: mime.getType(path) || 'application/octet-stream',
-                        ...getParams(key, params)
-                    }
-                }).promise();
-                promises.push(promise);
-                await promise;
-                spinner.text = chalk`Syncing...\n{dim   Uploaded {cyan ${key}}}`;
-            } catch (ex) {
-                spinner.fail(chalk`Upload failure for object {cyan ${key}}`);
-                console.error(pe.render(ex));
-                process.exit(1);
-            }
+            const task = {s3, publicDir, objects, isKeyInUse, path, config, params, spinner};
+            uploadQueue.push(task, (key?: string, err?: any) => {
+                if (!err) {
+                    if (key) spinner.text = chalk`Syncing...\n{dim   Uploaded {cyan ${key}}}`
+                } else {
+                    if (key) spinner.fail(chalk`Upload failure for file {cyan ${key}}`);
+                    console.error(pe.render(err));
+                    process.exit(1);
+                }
+            });
         });
 
         // now we play the waiting game.
         await streamToPromise(stream as any as Readable); // todo: find out why the typing won't allow this as-is
-        await Promise.all(promises);
-
-        if (config.removeNonexistentObjects) {
-            const objectsToRemove = objects.map(obj => ({ Key: <string>obj.Key })).filter(obj => obj.Key && !isKeyInUse[obj.Key]);
-
-            for (let i = 0; i < objectsToRemove.length; i += OBJECTS_TO_REMOVE_PER_REQUEST) {
-                const objectsToRemoveInThisRequest = objectsToRemove.slice(i, i + OBJECTS_TO_REMOVE_PER_REQUEST);
-
-                spinner.text = `Removing objects ${i + 1} to ${i + objectsToRemoveInThisRequest.length} of ${objectsToRemove.length}`;
-                await s3.deleteObjects({
-                    Bucket: config.bucketName,
-                    Delete: {
-                        Objects: objectsToRemoveInThisRequest,
-                        Quiet: true
-                    }
-                }).promise();
+        
+        uploadQueue.drain = async () => {
```

Comment:
Not a fan of using a callback here. The cleanest option would be to use parallelLimit instead, as it [supports promises/awaiting](https://github.com/caolan/async/blob/master/test/es2017/awaitableFunctions.js).

Another idea would be to make a helper method to convert drain into an awaitable, something like: (untested)

```typescript
const drainQueue(queue: QueueObject): Promise<void> => {
    return new Promise((resolve, reject) => {
        queue.drain = resolve;
        queue.error = (err, task) => {
            queue.kill();
            reject(err); // If you want to get really fancy, make a custom error object that includes task in it
        };
    });
}
```